amida-tech/indaba-client
# Task/inba 50 survey

#### What's this PR do?
Gives the test admin the ability to create their own surveys. This page is not complete, but does handle the lion's share of the functionality we needed. 

I apologize that this is huge. I shouldn't have let this go as long as it did, but I had my nose to the grindstone. I find it easier to edit when the manuscript is mostly complete. Tired. It's hard to think right now.

#### Related JIRA tickets:
https://jira.amida-tech.com/browse/INBA-521 Rewrite in Redux for the future. 

#### How should this be manually tested?
**Note:** If a bug is discovered (and I'm sure there will be), create a JIRA ticket as long as it does not impede access to other pages. If the Survey Builder _does_ prevent access to pages, reject the PR until it is fixed. 

Load and login as the test admin. Go to either the Create Project Wizard or the Project Management (testing both is best but they use the same page). Create a survey by entering a name into the survey creation field. Click on the "Survey" tab.

Enter instructions and a section name for the first section. Add a new section. 
Click on a "+" icon for the new questions on the left. If the "View All Questions" view dropdown option is selected, you will be prompted for which section you would like to add it to. Select one and ensure that the question appears under that section. 

Click the view dropdown. Ensure that any unnamed fields have a default value of "Section X" and named sections have the actual name. Select one.

Ensure that the view beneath it rerenders to display only that section. 

Among your questions, try the arrow buttons and delete button. Arrow buttons should move the question up and down the section (and explaining why the user cannot move the question further). Delete a question. 

Enter values into a question. Click on the "Save Progress" button. Ensure that values are entered into the `surveyService`. Reload the page and ensure that the data repopulates the survey builder.

**Note:** You can also assign a task, log out and log into that user. On the user's tasks, the survey questions will appear. However, there is a bug or issue where the changing the answers fails to update due to an error. I'm speaking to Steve and Afsin about this to fix. 

**Note:** I would recommend avoiding Bullet Point, Drop Downs and Scale questions for now as I'm not certain survey services is really supporting those yet. If you add one, you can still delete it. 

#### Any background context you want to provide?
Requires validation (ensure that all questions have a `text` field and all questions with `choices` have at least two choices to choose from. Believe it or not, you can _still_ save questions without the text field or choices. This is kind of a problem.

The "Expand" and "Collapse" buttons don't do anything at the moment because they're supposed to work with an accordion. I might want to talk management into removing this just to ease this.

I _hate_ the grommet menu dropdown. It's badly messing up styling. In hindsight, I should have probably removed the icon and used it off the whole field. But Grommet dropdown's still have that issue where it sticks where it is on the screen when you scroll. Lacey said we needed to update grommet to fix that. Or make our own.

Need to talk about the important of weighted questions.

Make editing all this unavailable when the survey is published.

Requires existing questions.
#### Screenshots (if appropriate):
